### PR TITLE
Strong detection for List/Dictionary concurrent use in Debug

### PR DESCRIPTION
**Pull request by benaadams**
Bit ugly, but should do the trick?

Resolves: https://github.com/dotnet/coreclr/issues/17070

### CODE REVIEW

**Comment by benaadams** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs [hunk 1]
```diff
@@ -1041,6 +1191,9 @@ ICollection IDictionary.Values
                 {
                     ThrowHelper.ThrowWrongKeyTypeArgumentException(key, typeof(TKey));
                 }
+#if DEBUG
```
Comment (sharwell) [inline src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: ❓ Why bother with `#if DEBUG` here if the target method is marked `[Conditional("DEBUG")]`?

Comment (danmosemsft) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: If you do that it might be worth naming something like as DebugOnlyConcurrentAccessCheck so it's obvious it's debug only

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs [hunk 2]
```diff
@@ -214,14 +215,26 @@ public ValueCollection Values
         {
             get
             {
+#if DEBUG
+                int version = _version;
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: if you're using the existing `_version` field doesn't it have to be volatile or incremented with Interlocked?

@stephentoub

Comment (benaadams) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: In `DEBUG` its incremented in `ConcurrentAccessWriteCheck` with
```csharp
Interlocked.CompareExchange(ref _version, version + 1, version) != version
```

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs (hunk 1 above)
Comment (benaadams) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: Wasn't sure the Jit would elide all the `int version = _version;` calls (or C# delete them and not produce an error about not using it)

Comment (benaadams) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: Will change an check diffs as it reduces the number of `#if`s considerably

**Comment by benaadams** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by jkotas** [comment]
The barriers from these interlocked operations can hide some types of the races as well.

Do we have any examples of bugs in CoreFX that were caused by invalid multi-threaded Dictionary use and that were in the product for a long time because of they were hard to hit and find? It would be nice to validate the effectiveness of this on some real world examples.

**Comment by benaadams** [comment]
Its generating the asm for the `[Conditional("DEBUG")]` in Release; though not calling it, also not eliding the `_version` use :sad: back to `#if DEBUG`
```
Total bytes of diff: 1604 (0.05% of base)
    diff is a regression.

Total byte diff includes 61 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    2 unique methods,       61 unique bytes

Top file regressions by size (bytes):
        1604 : System.Private.CoreLib.dasm (0.05% of base)

1 total files with size differences (0 improved, 1 regressed), 0 unchanged.

Top method regessions by size (bytes):
          88 : System.Private.CoreLib.dasm - Dictionary`2:.ctor(ref,ref):this (58 methods)
          87 : System.Private.CoreLib.dasm - Dictionary`2:CopyTo(ref,int):this (29 methods)
          87 : System.Private.CoreLib.dasm - Dictionary`2:System.Collections.IDictionary.get_Item(ref):ref:this (29 methods)
          75 : System.Private.CoreLib.dasm - Dictionary`2:TrimExcess(int):this (25 methods)
          58 : System.Private.CoreLib.dasm - Dictionary`2:Resize(int,bool):this (29 methods)

Top method improvements by size (bytes):
        -203 : System.Private.CoreLib.dasm - Dictionary`2:Clear():this (29 methods)

164 total methods with size differences (1 improved, 163 regressed), 17109 unchanged.
```

**Comment by benaadams** [comment]
```
Total bytes of diff: -66 (0.00% of base)
    diff is an improvement.

Total byte diff includes 0 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    0 unique methods,        0 unique bytes

Top file improvements by size (bytes):
         -66 : System.Private.CoreLib.dasm (0.00% of base)

1 total files with size differences (1 improved, 0 regressed), 0 unchanged.

Top method regessions by size (bytes):
          75 : System.Private.CoreLib.dasm - Dictionary`2:TrimExcess(int):this (25 methods)
          58 : System.Private.CoreLib.dasm - Dictionary`2:Resize(int,bool):this (29 methods)
          32 : System.Private.CoreLib.dasm - PseudoCustomAttribute:GetCustomAttributes(ref,ref,byref):ref (9 methods)
          32 : System.Private.CoreLib.dasm - PseudoCustomAttribute:IsDefined(ref,ref):bool (9 methods)
           8 : System.Private.CoreLib.dasm - Attribute:CopyToArrayList(ref,ref,ref)
           8 : System.Private.CoreLib.dasm - MemberInfoCache`1:PopulateInterfaces(struct):ref:this
           8 : System.Private.CoreLib.dasm - MemberInfoCache`1:PopulateEvents(struct,ref,ref,byref):this
           8 : System.Private.CoreLib.dasm - EventSource:DebugCheckEvent(byref,ref,ref,ref,ref,int)
           8 : System.Private.CoreLib.dasm - ManifestBuilder:AddEventParameter(ref,ref):this
           8 : System.Private.CoreLib.dasm - SerializationInfo:AddValueInternal(ref,ref,ref):this
           8 : System.Private.CoreLib.dasm - EventRegistrationTokenTable`1:AddEventHandlerNoLock(ref):struct:this

Top method improvements by size (bytes):
        -319 : System.Private.CoreLib.dasm - Dictionary`2:Clear():this (29 methods)

12 total methods with size differences (1 improved, 11 regressed), 17261 unchanged.
```

Regression in Dictionary is from incrementing `_version` in `TrimExcess` and `Resize`

Other regressions are from `ContainsKey(TKey key)` deciding to inline now that it gets the `bool` to a local and returns it; rather than returning the `bool` directly - which seems a bit weird?

**Comment by danmosemsft** [comment]
@jkotas this was motivated by https://github.com/dotnet/corefx/issues/28198. I assume there have been other cases periodically?

**Comment by benaadams** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by benaadams** [comment]
Is there a way to run coreclr (checked or debug) against corefx outerloop?

**Comment by benaadams** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs (hunk 2 above)
Comment (benaadams) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: Changed to
```csharp
if (Interlocked.Exchange(ref _version, version + 1) != version)
{
    ThrowHelper.ThrowInvalidOperationException_ConcurrentOperationsNotSupported();
}
```
As don't care if it stops version being updated; just that it throws if the version wasn't correct

**Comment by benaadams** [comment]
Triggered some tests that were checking version didn't change
```
System.Collections.Tests.Dictionary_Generic_Tests_string_string.EnsureCapacity_Generic_RequestingLargerCapacity_DoesNotInvalidateEnumeration(count: 1) [FAIL]
   System.InvalidOperationException : Collection was modified; enumeration operation may not execute.
   Stack Trace:
         at System.Collections.Generic.Dictionary`2.KeyCollection.Enumerator.MoveNext()
      D:\j\workspace\x64_checked_w---d7295605\_\fx\src\System.Collections\tests\Generic\Dictionary\Dictionary.Generic.Tests.netcoreapp.cs(104,0): at System.Collections.Tests.Dictionary_Generic_Tests`2.EnsureCapacity_Generic_RequestingLargerCapacity_DoesNotInvalidateEnumeration(Int32 count)
System.Collections.Tests.Dictionary_Generic_Tests_string_string.EnsureCapacity_Generic_RequestingLargerCapacity_DoesNotInvalidateEnumeration(count: 75) [FAIL]
   System.InvalidOperationException : Collection was modified; enumeration operation may not execute.
   Stack Trace:
         at System.Collections.Generic.Dictionary`2.KeyCollection.Enumerator.MoveNext()
      D:\j\workspace\x64_checked_w---d7295605\_\fx\src\System.Collections\tests\Generic\Dictionary\Dictionary.Generic.Tests.netcoreapp.cs(104,0): at System.Collections.Tests.Dictionary_Generic_Tests`2.EnsureCapacity_Generic_RequestingLargerCapacity_DoesNotInvalidateEnumeration(Int32 count)
```

**Comment by benaadams** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by jkotas** [comment]
> @jkotas this was motivated by dotnet/corefx#28198

https://github.com/dotnet/corefx/issues/28198#issuecomment-374676611 says that we do not have any tests for the buggy code. If it is the case, this would not help us to find this issue faster.

**Comment by benaadams** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by benaadams** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by benaadams** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by benaadams** [comment]
Linux - CoreFx issues unrelated
```
Test Result (4 failures / ±0)
System.IO.Tests.FileInfo_GetSetTimes.TimesIncludeMillisecondPart_Linux
System.IO.Tests.File_GetSetTimes.TimesIncludeMillisecondPart_Linux
System.IO.Tests.DirectoryInfo_GetSetTimes.TimesIncludeMillisecondPart_Linux
System.IO.Tests.Directory_GetSetTimes.TimesIncludeMillisecondPart_Linux
```

**Comment by danmosemsft** [comment]
@stephentoub thoughts about existence of real world examples?

**Comment by benaadams** [comment]
@danmosemsft it didn't catch anything 😢 

Is this change worth it?

**Comment by danmosemsft** [comment]
I figured stress runs would be more likely to catch something. As to whether it's worth it, I'm open minded, I'm curious whether @stephentoub has seen real cases.

Do you understand this gain - is it worth keeping?
```
Top method improvements by size (bytes):
        -319 : System.Private.CoreLib.dasm - Dictionary`2:Clear():this (29 methods)
```

**Comment by benaadams** [comment]
> Do you understand this gain - is it worth keeping?

Added PR for the gains :) https://github.com/dotnet/coreclr/pull/17096

**Comment by stephentoub** [comment]
> I'm curious whether @stephentoub has seen real cases.

I've seen real cases, e.g. the SQL one I fixed earlier in the week, but as Jan restated this wouldn't have caught that as we didn't have any tests that exercised those code paths, nevermind ones that would have exercised it in parallel.

I just did a search for "static Dictionary", "static readonly Dictionary", "readonly static Dictionary" in corefx (thankfully there's only one of the latter, as that's not the desired ordering), and did a cursory review of each's usage.  A few things I noticed:
- This is buggy: https://github.com/dotnet/corefx/blob/619dbe6503a2de2afa27122cb34716d9ff6906ff/src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/DataContractResolverLibrary.cs#L16-L34.  But it's in tests, and it would only manifest if tests that were using this type were to run concurrently; it looks like all tests that use it are in the same xunit class, so they won't run concurrently.  Still, it's probably worth fixing just to avoid issues in the future should usage here be extended.
- Just based on brief code inspection, these look problematic: https://github.com/dotnet/corefx/blob/619dbe6503a2de2afa27122cb34716d9ff6906ff/src/Common/tests/System/Runtime/Serialization/Utils.cs#L38-L39.  Those dictionaries are read/written without locks from various static helper methods.  These could be in the same camp as above, though, where they're "safe" only because all tests that use them end up getting serialized.
- In the same boat: https://github.com/dotnet/corefx/blob/619dbe6503a2de2afa27122cb34716d9ff6906ff/src/System.ComponentModel.Composition/tests/System/ComponentModel/Composition/DynamicMetadata.cs#L144.  It's accessed in an unsafe way, but it appears to be ok because the tests that use it are serialized by xunit.  If we were to ever refactor these tests to live in different test classes, this could become a test bug.
- There's a potential product issue here, though not one these changes would likely catch: https://github.com/dotnet/corefx/blob/619dbe6503a2de2afa27122cb34716d9ff6906ff/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnectionString.cs#L488-L492. cc: @saurabh500. There's no synchronization whatsoever (no locks, no interlockeds, no volatiles, etc.) used in lazily initializing this dictionary.  It's possible that could be an issue on some hardware.
- This looks dangerous: https://github.com/dotnet/corefx/blob/bb399884797d73f5401c637ed861745b97f0fd92/src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContract.cs#L30-L36.  But I also can't find any references to GetDataContracts() elsewhere in corefx, so usage isn't clear.  And enough care was taken with other shared dictionaries elsewhere in the file that presumably the author here knew what they were doing.
- This looks like a product bug, and one that this change could potentially catch (cc: @shmao). https://github.com/dotnet/corefx/blob/bb399884797d73f5401c637ed861745b97f0fd92/src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ExtensionDataReader.cs#L487-L496.  But again I don't have enough context as to usage, and I don't know if we have sufficient tests in place to help this change catch it.  Similarly it's suspicious that there are multiple static methods that access another dictionary on this type, writing https://github.com/dotnet/corefx/blob/bb399884797d73f5401c637ed861745b97f0fd92/src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ExtensionDataReader.cs#L501-L505 and reading https://github.com/dotnet/corefx/blob/bb399884797d73f5401c637ed861745b97f0fd92/src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ExtensionDataReader.cs#L209 and without explicit synchronization, at least at this level.

In short, across the large codebase that is corefx, it's possible this might catch something, but I wouldn't expect significant wins.

**Comment by jkotas** [comment]
I have opened https://github.com/dotnet/corefx/issues/30651 and https://github.com/dotnet/corefx/issues/30650 on the issues that Stephen identified by code inspection.

**Comment by jkotas** [comment]
Otherwise, I do not think that this is worth the extra DEBUG #ifdefs. I think that these ifdefs have about the same chance of catching problems when running debug CoreCLR build as introducing new ones.

@benaadams Thanks for the suggested change anyway.
